### PR TITLE
chroot: add apply_, sls and highstate for state execution

### PR DESCRIPTION
**Pull request by aplanas**
### What does this PR do?

Based on the code from the SSH client wrapper, add apply_, sls and highstate for the execution of Salt states inside a chroot environment.

### Tests written?

Yes

### CODE REVIEW

**Comment by aplanas** [comment]
This commit is on top of https://github.com/saltstack/salt/pull/54959. If this gets merged, I will rebase this one.

**Comment by Akm0d** [comment]
re-run full pycryptodomex

**Comment by Akm0d** [comment]
re-run full pycryptodomex

File: tests/unit/modules/test_chroot.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                '/chroot',
+                ['python{}'.format(sys.version_info[0]), '/tmp01/salt-call',
+                 '--metadata', '--local',
+                 '--log-file', '/tmp01/log', '--cachedir', '/tmp01/cache',
+                 '--out', 'json', '-l', 'quiet', '--', 'test.ping'])
+            utils_mock['files.rm_rf'].assert_called_once()
+
+    @patch('salt.modules.chroot.exist')
+    @patch('tempfile.mkdtemp')
+    def test_call_success_parameters(self, mkdtemp, exist):
```
Comment (Akm0d) [inline tests/unit/modules/test_chroot.py]: This test is failing on some OSes:
```Traceback (most recent call last):
  File "/tmp/kitchen/testing/.nox/runtests-parametrized-3-coverage-true-crypto-none-transport-zeromq/lib/python3.5/site-packages/mock/mock.py", line 1330, in patched
    return func(*args, **keywargs)
  File "/tmp/kitchen/testing/tests/unit/modules/test_chroot.py", line 236, in test_call_success_parameters
    '--', 'ssh.set_auth_key', 'user=user', 'key=key'])
  File "/tmp/kitchen/testing/.nox/runtests-parametrized-3-coverage-true-crypto-none-transport-zeromq/lib/python3.5/site-packages/mock/mock.py", line 944, in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
  File "<string>", line 3, in raise_from
AssertionError: expected call not found.
Expected: mock('/chroot', ['python3', '/tmp01/salt-call', '--metadata', '--local', '--log-file', '/tmp01/log', '--cachedir', '/tmp01/cache', '--out', 'json', '-l', 'quiet', '--', 'ssh.set_auth_key', 'user=user', 'key=key'])
Actual: mock('/chroot', ['python3', '/tmp01/salt-call', '--metadata', '--local', '--log-file', '/tmp01/log', '--cachedir', '/tmp01/cache', '--out', 'json', '-l', 'quiet', '--', 'ssh.set_auth_key', 'key=key', 'user=user'])
```

File: tests/unit/modules/test_chroot.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                'comment': 'Error'
+            })
+            utils_mock['thin.gen_thin'].assert_called_once()
+            salt_mock['config.option'].assert_called()
+            salt_mock['cmd.run'].assert_called_once()
+            utils_mock['files.rm_rf'].assert_called_once()
+
+    @patch('salt.modules.chroot.exist')
+    @patch('tempfile.mkdtemp')
+    def test_call_fails_salt_thin(self, mkdtemp, exist):
```
Comment (Akm0d) [inline tests/unit/modules/test_chroot.py]: This test is failing on windows: 
```Traceback (most recent call last):
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\.nox\runtests-parametrized-2-7-coverage-true-crypto-none-transport-zeromq\lib\site-packages\mock\mock.py", line 1330, in patched
    return func(*args, **keywargs)
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\tests\unit\modules\test_chroot.py", line 162, in test_call_fails_salt_thin
    '--out', 'json', '-l', 'quiet', '--', 'test.ping'])
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\.nox\runtests-parametrized-2-7-coverage-true-crypto-none-transport-zeromq\lib\site-packages\mock\mock.py", line 944, in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\.nox\runtests-parametrized-2-7-coverage-true-crypto-none-transport-zeromq\lib\site-packages\six.py", line 737, in raise_from
    raise value
AssertionError: expected call not found.
Expected: mock(u'/chroot', [u'python2', u'/tmp01/salt-call', u'--metadata', u'--local', u'--log-file', u'/tmp01/log', u'--cachedir', u'/tmp01/cache', u'--out', u'json', u'-l', u'quiet', u'--', u'test.ping'])
Actual: mock(u'/chroot', ['python2', '\\tmp01\\salt-call', '--metadata', '--local', '--log-file', '\\tmp01\\log', '--cachedir', '\\tmp01\\cache', '--out', 'json', '-l', 'quiet', '--', 'test.ping'])
```

File: tests/unit/modules/test_chroot.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                '/chroot',
+                ['python{}'.format(sys.version_info[0]), '/tmp01/salt-call',
+                 '--metadata', '--local',
+                 '--log-file', '/tmp01/log', '--cachedir', '/tmp01/cache',
+                 '--out', 'json', '-l', 'quiet', '--', 'test.ping'])
+            utils_mock['files.rm_rf'].assert_called_once()
+
+    @patch('salt.modules.chroot.exist')
+    @patch('tempfile.mkdtemp')
+    def test_call_success(self, mkdtemp, exist):
```
Comment (Akm0d) [inline tests/unit/modules/test_chroot.py]: This test is failing on windows:
```Traceback (most recent call last):
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\.nox\runtests-parametrized-2-7-coverage-true-crypto-none-transport-zeromq\lib\site-packages\mock\mock.py", line 1330, in patched
    return func(*args, **keywargs)
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\tests\unit\modules\test_chroot.py", line 198, in test_call_success
    '--out', 'json', '-l', 'quiet', '--', 'test.ping'])
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\.nox\runtests-parametrized-2-7-coverage-true-crypto-none-transport-zeromq\lib\site-packages\mock\mock.py", line 944, in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
  File "c:\users\admini~1\appdata\local\temp\kitchen\testing\.nox\runtests-parametrized-2-7-coverage-true-crypto-none-transport-zeromq\lib\site-packages\six.py", line 737, in raise_from
    raise value
AssertionError: expected call not found.
Expected: mock(u'/chroot', [u'python2', u'/tmp01/salt-call', u'--metadata', u'--local', u'--log-file', u'/tmp01/log', u'--cachedir', u'/tmp01/cache', u'--out', u'json', u'-l', u'quiet', u'--', u'test.ping'])
Actual: mock(u'/chroot', ['python2', '\\tmp01\\salt-call', '--metadata', '--local', '--log-file', '\\tmp01\\log', '--cachedir', '\\tmp01\\cache', '--out', 'json', '-l', 'quiet', '--', 'test.ping'])
```

Comment (Akm0d) [reply tests/unit/modules/test_chroot.py]: I don't imagine the `chroot` module works on windows so we can probably just add a `@skipIf(salt.utils.platform.is_windows())` to the top of the test class.

**Comment by Akm0d** [comment]
I don't imagine the chroot module works on windows so we can probably just add a @skipIf(salt.utils.platform.is_windows()) to the top of the test class.

File: tests/unit/modules/test_chroot.py (hunk 1 above)
Comment (Akm0d) [reply tests/unit/modules/test_chroot.py]: It looks like `key=key` and `user=user` changed places on the actual return

Comment (aplanas) [reply tests/unit/modules/test_chroot.py]: Ah this is tricky, as the ordering of kwargs makes the ordering unpredictable. This tests comes from https://github.com/saltstack/salt/pull/54959. I will figure a fix there and rebase this commit on top of the other one.

Thanks a lot!

**Comment by aplanas** [comment]
> I don't imagine the `chroot` module works on windows so we can probably just add a `@skipIf(salt.utils.platform.is_windows())` to the top of the test class.

I will do this. Thanks for the reviews!!!!

File: tests/unit/modules/test_chroot.py (hunk 3 above)
Comment (aplanas) [reply tests/unit/modules/test_chroot.py]: Done

File: tests/unit/modules/test_chroot.py (hunk 2 above)
Comment (aplanas) [reply tests/unit/modules/test_chroot.py]: Done

File: tests/unit/modules/test_chroot.py (hunk 1 above)
Comment (aplanas) [reply tests/unit/modules/test_chroot.py]: Done

**Review by dwoz — APPROVED** [review]

**Comment by aplanas** [comment]
Thanks!

**Pull request merged**